runtime/controller.go

```diff
@@ -118,6 +126,165 @@ func NewController(ctx context.Context, rt *Runtime, instanceID string, logger *
+			Level:   slogLevelToPBLevel(record.Level),
+			Time:    timestamppb.New(record.Time),
+			Message: record.Message,
+			Payload: payload,
+			Size:    uint32(size),
+		}
+		b.broadcast(message)
+	}()
+	return b.zapHandler.Handle(ctx, record)
+}
+
+func slogLevelToPBLevel(level slog.Level) runtimev1.LogLevel {
+	switch level {
+	case -4:
+		return runtimev1.LogLevel_LOG_LEVEL_DEBUG
+	case 0:
+		return runtimev1.LogLevel_LOG_LEVEL_INFO
+	case 4:
+		return runtimev1.LogLevel_LOG_LEVEL_WARN
+	case 8:
+		return runtimev1.LogLevel_LOG_LEVEL_ERROR
+	default:
+		return runtimev1.LogLevel_LOG_LEVEL_UNSPECIFIED
+	}
+}
+
+func (b *broadcastHandler) WithAttrs(attrs []slog.Attr) slog.Handler {
+	return b.zapHandler.WithAttrs(attrs)
+}
+
+func (b *broadcastHandler) WithGroup(name string) slog.Handler {
+	return b.zapHandler.WithGroup(name)
+}
+
+func (b *broadcastHandler) addClient(client chan *runtimev1.Log, replay bool) {
+	b.mu.Lock()
+	b.clients[client] = struct{}{}
+	messages := append([]*runtimev1.Log(nil), b.messages...)
+	b.mu.Unlock()
+
+	if replay {
+		// Send historical messages in a non-blocking way.
+		go func() {
+			for _, msg := range messages {
+				client <- msg
+			}
+		}()
+	}
+}
+
+func (b *broadcastHandler) removeClient(client chan *runtimev1.Log) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+	delete(b.clients, client)
+	close(client) // close the channel here after the client is removed.
+}
+
+func (b *broadcastHandler) broadcast(message *runtimev1.Log) {
+	b.mu.Lock()
+	defer b.mu.Unlock()
+
+	b.bufferSize += int64(message.Size)
+	b.messages = append(b.messages, message)
+
+	// Trim the old messages if buffer size exceeds the limit.
+	for b.bufferSize > b.maxBufferSize && len(b.messages) > 0 {
+		removedMsg := b.messages[0]
+		b.messages = b.messages[1:]
+		b.bufferSize -= int64(removedMsg.Size)
+	}
+
+	for client := range b.clients {
+		client <- message
+	}
+}
+
+func (c *Controller) WatchLogs(srv runtimev1.RuntimeService_WatchLogsServer, replay bool) error {
+	messageChannel := make(chan *runtimev1.Log)
+	c.broadcastHandler.addClient(messageChannel, replay)
+	defer c.broadcastHandler.removeClient(messageChannel)
+
+	for {
+		select {
+		case message, open := <-messageChannel:
+			if !open {
+				panic("client closed!")
+			}
+			err := srv.Send(&runtimev1.WatchLogsResponse{
+				Log: message,
+			})
+			if err != nil {
+				return err
+			}
+		case <-srv.Context().Done():
+			return nil
+		}
+	}
+}
+
+func (c *Controller) GetLogs(asc bool) []*runtimev1.Log {
+	c.broadcastHandler.mu.Lock()
+	messages := append([]*runtimev1.Log(nil), c.broadcastHandler.messages...)
+	c.broadcastHandler.mu.Unlock()
+	var startIdx, endIdx int
+	if asc {
+		startIdx = 0
+		endIdx = len(messages)
+	} else {
+		startIdx = len(messages) - 1
+		endIdx = -1
+	}
+	logs := make([]*runtimev1.Log, 0, len(messages))
+	for i := startIdx; i != endIdx; {
+		logs = append(logs, messages[i])
+		if asc {
+			i++
+		} else {
+			i--
+		}
+	}
+	return logs
+}
```

Comment:
Some general comments:
1. It would be good to avoid passing the API signature types into the controller – instead, it could take a normal callback function, which the server wraps the API handler with. Take a look at the `ListResources` and `WatchResources` APIs in the async reconcile PR: https://github.com/rilldata/rill/pull/3092
2. Also consider implementing the `replay` in the API handler by calling `GetLogs` before starting a subscription, instead of propagating it here (perfect consistency for replay isn't needed). Again, can look at `WatchResources` for inspiration.
3. Should we also support a `page_size` and `page_token` for `GetLogs`? And maybe limiting replay? I'm thinking about the UX for showing a logs screen on the status page – might be a bit heavy to fetch 16MB logs at the start each time.